File: chapter_2_linked_lists/python/2.1-remove_dups.py
SinglyLinkedList = __import__('singly_linked_list').SinglyLinkedList


class SinglyLinkedList_2_1(SinglyLinkedList):

    def del_dups(self):
        """
        Delete nodes with duplicate values in a linked list
        O(n) time complexity
        """
        curr = self.head
        _set = set()
        index = 0
        while curr:
            if curr.value in _set:
                self.delete_node(index)
            else:
                _set.add(curr.value)
                index += 1
            curr = curr.next


sll = SinglyLinkedList_2_1()
sll.add_to_front(1)
sll.add_to_front(1)
sll.add_to_front(2)
sll.add_to_front(4)
sll.add_to_front(4)
sll.add_to_front(8)
sll.add_to_front(8)

sll.print_linked_list()  # 8-> 8-> 4-> 4-> 2-> 1-> 1
sll.del_dups()
sll.print_linked_list()  # 8-> 4-> 2-> 1

File: chapter_2_linked_lists/js/2.1-removeDups.js
const { SinglyLinkedList } = require("./singlyLinkedList")

class SinglyLinkedList_2_1 extends SinglyLinkedList {
  /**
   * Delete nodes with duplicate values in a linked list.
   * O(n) time complexity.
   */
  removeDups() {
    const set = new Set()
    let curr = this.head
    let index = 0
    while (curr) {
      const val = curr.value
      if (set.has(val)) {
        this.deleteNode(index)
      } else {
        set.add(val)
        index++
      }
      curr = curr.next
    }
  }
}

const sll = new SinglyLinkedList_2_1()
const arr = [1, 1, 2, 2, 4, 8, 16, 16]
for (let i of arr) sll.addNode(i)
sll.printList() // 16 -> 16 -> 8 -> 4 -> 2 -> 2 -> 1 -> 1
sll.removeDups()
sll.printList() // 16 -> 8 -> 4 -> 2 -> 1
